python_modules/dagster/dagster/core/execution_plan/intermediates_manager.py

```diff
@@ -0,0 +1,115 @@
+from abc import ABCMeta, abstractmethod
+from collections import namedtuple
+import os
+import pickle
+
+import six
+
+from dagster import check
+from dagster.utils import mkdir_p
+
+
+class StepOutputHandle(namedtuple('_StepOutputHandle', 'step_key output_name')):
+    @staticmethod
+    def from_step(step, output_name):
+        from .objects import ExecutionStep
+
+        check.inst_param(step, 'step', ExecutionStep)
+
+        return StepOutputHandle(step.key, output_name)
+
+    def __new__(cls, step_key, output_name):
+        return super(StepOutputHandle, cls).__new__(
+            cls,
+            step_key=check.str_param(step_key, 'step_key'),
+            output_name=check.str_param(output_name, 'output_name'),
+        )
+
+
+def read_pickle_file(path):
+    check.str_param(path, 'path')
+    with open(path, 'rb') as ff:
+        return pickle.load(ff)
```

Comment:
fwiw, this will be an issue if we somehow have py2 and py3 processes attempting to read the same pickled intermediate values. it might be pedantic, but this will be an issue even if we set a least-commmon-denominator pickle protocol (i.e., 2), because of unicode handling. we might want to do something to fail hard if we try to read and write to the same on-disk intermediate file cache from processes running incompatible python versions.

Reply:
what is the best way to deal with this issue?

Reply:
we could prefix/suffix the pickle filenames with a tag to mark which python version created them, and bail hard if attempting to read from an incompatible version?

Reply:
https://docs.python.org/2/library/tempfile.html#tempfile.gettempdir -- but this gives weird and undesirable results on os x